File: 프로그래머스/2/12981. 영어 끝말잇기/영어 끝말잇기.java
import java.util.*;
class Solution {
    public int[] solution(int n, String[] words) {
        int[] answer = {0,0};
        int lastWordIdx = 0;
        List<String> wordList = new ArrayList<>();
        wordList.add(words[0]);
        for(int i=1;i<words.length;i++) {
            String preWord = wordList.get(i-1);
            if(wordList.contains(words[i]) || 
              !(preWord.substring(preWord.length()-1)
               .equals(words[i].substring(0,1)))) {
                lastWordIdx = i;
                break;
            }
            wordList.add(words[i]);
            if(wordList.size()==words.length)
                return answer;
        }
        if((lastWordIdx+1) % n == 0) {
            answer[0] = n;
            answer[1] = (lastWordIdx+1) / n;
        }
        else{
            answer[0] = (lastWordIdx+1) % n;
            answer[1] = (lastWordIdx+1) / n +1;
        }
        return answer;
    }
}

File: 프로그래머스/2/12981. 영어 끝말잇기/영어 끝말잇기.js
function solution(n, words) {
    var answer = [];
    const dup = new Set()//중복 체크용 set
    dup.add(words[0])
    let preLen = 1 
    for(let i=1;i<=words.length-1;i++) {
        dup.add(words[i])
        
        if(preLen >= dup.size) { // 중복이면
            //i번째 값이 몇 번째 번호인지, 몇 번째 사이클인지
            answer.push(i%n+1)
            answer.push(Math.ceil((i+1)/n))
            return answer
        }
        preLen++
        // console.log(preLen,dup.size)
        let el1 = words[i-1].substring(words[i-1].length-1)
        let el2 = words[i].substring(0,1)
        // console.log(el1, el2)
        if(el1!==el2) {
            answer.push(i%n+1)
            answer.push(Math.ceil((i+1)/n))
            return answer
        }
    }
    return [0,0]
}